Sysbench read and write workload hit 9007 Write conflict error

## Bug Report

<!-- Thanks for your bug report! Don't worry if you can't fill out all the sections. -->

### What version of TiKV are you using?
<!-- You can run `tikv-server --version` -->
/ # /tikv-server -V
TiKV
Release Version:   5.3.0
Edition:           Community
Git Commit Hash:   6c1424706f3d5885faa668233f34c9f178302f36
Git Commit Branch: heads/refs/tags/v5.3.0
UTC Build Time:    2021-11-22 10:51:38
Rust Version:      rustc 1.56.0-nightly (2faabf579 2021-07-27)
Enable Features:   jemalloc mem-profiling portable sse protobuf-codec test-engines-rocksdb cloud-aws cloud-gcp
Profile:           dist_release

### What operating system and CPU are you using?
<!-- If you're using Linux, you can run `cat /proc/cpuinfo` -->

### Steps to reproduce
<!-- If possible, provide a recipe for reproducing the error. A complete runnable program is good. -->
run oltp test case oltp_stability_002 failed.

### What did you expect?
Case run succeeded.

### What did happened?
Hit following error.
2021-12-08T12:03:47.527+0800	INFO	oltppkg/workload.go:331	FATAL: mysql_stmt_execute() returned error 9007 (Write conflict, txnStartTS=429637298355962059, conflictStartTS=429637298355962016, conflictCommitTS=429637298382176376, key={tableID=156, indexID=1, indexValues={5025659, 5045579, }} primary={tableID=156, handle=5045579} [try again later]) for query 'COMMIT'

/var/lib/tidb/log # grep -A5 -B 5 '429637298355962016' *.log
tidb.log-[2021/12/08 04:03:47.090 +00:00] [WARN] [pd.go:152] ["get timestamp too slow"] ["cost time"=31.14879ms]
tidb.log-[2021/12/08 04:03:47.265 +00:00] [ERROR] [terror.go:307] ["encountered error"] [error=EOF] [stack="github.com/pingcap/tidb/parser/terror.Log\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/parser/terror/terror.go:307\ngithub.com/pingcap/tidb/server.(*Server).onConn\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/server/server.go:519"]
tidb.log-[2021/12/08 04:03:47.523 +00:00] [WARN] [client_batch.go:794] ["wait response is cancelled"] [to=tc-tikv-2.tc-tikv-peer.testbed-oltp-hm-k52vt.svc:20160] [cause="context canceled"]
tidb.log-[2021/12/08 04:03:47.524 +00:00] [WARN] [client_batch.go:794] ["wait response is cancelled"] [to=tc-tikv-1.tc-tikv-peer.testbed-oltp-hm-k52vt.svc:20160] [cause="context canceled"]
tidb.log-[2021/12/08 04:03:47.524 +00:00] [WARN] [client_batch.go:794] ["wait response is cancelled"] [to=tc-tikv-3.tc-tikv-peer.testbed-oltp-hm-k52vt.svc:20160] [cause="context canceled"]
tidb.log:[2021/12/08 04:03:47.524 +00:00] [WARN] [session.go:721] ["can not retry txn"] [conn=29083] [label=general] [error="[kv:9007]Write conflict, txnStartTS=429637298355962059, conflictStartTS=429637298355962016, conflictCommitTS=429637298382176376, key={tableID=156, indexID=1, indexValues={5025659, 5045579, }} primary={tableID=156, handle=5045579} [try again later]"] [IsBatchInsert=false] [IsPessimistic=true] [InRestrictedSQL=false] [tidb_retry_limit=10] [tidb_disable_txn_auto_retry=true]
tidb.log:[2021/12/08 04:03:47.524 +00:00] [WARN] [session.go:737] ["commit failed"] [conn=29083] ["finished txn"="Txn{state=invalid}"] [error="[kv:9007]Write conflict, txnStartTS=429637298355962059, conflictStartTS=429637298355962016, conflictCommitTS=429637298382176376, key={tableID=156, indexID=1, indexValues={5025659, 5045579, }} primary={tableID=156, handle=5045579} [try again later]"]
tidb.log:[2021/12/08 04:03:47.524 +00:00] [INFO] [conn.go:1069] ["command dispatched failed"] [conn=29083] [connInfo="id:29083, addr:10.233.116.131:46984 status:10, collation:utf8mb4_general_ci, user:root"] [command=Execute] [status="inTxn:0, autocommit:1"] [sql=COMMIT] [txn_mode=PESSIMISTIC] [err="[kv:9007]Write conflict, txnStartTS=429637298355962059, conflictStartTS=429637298355962016, conflictCommitTS=429637298382176376, key={tableID=156, indexID=1, indexValues={5025659, 5045579, }} primary={tableID=156, handle=5045579} [try again later]\nCOMMIT"]


Logs and monitor can be get from minio via testbed testbed-oltp-hm-k52vt.

--- a/src/storage/txn/actions/prewrite.rs
+++ b/src/storage/txn/actions/prewrite.rs
@@ -342,12 +342,25 @@ impl<'a> PrewriteMutation<'a> {
     ) -> Result<Option<(Write, TimeStamp)>> {
         match reader.seek_write(&self.key, TimeStamp::max())? {
             Some((commit_ts, write)) => {
-                // Abort on writes after our start timestamp ...
-                // If exists a commit version whose commit timestamp is larger than current start
+                // Abort on writes after our start/for_update timestamp ...
+                // If exists a commit version whose commit timestamp is larger than current start/for_update
                 // timestamp, we should abort current prewrite.
-                if commit_ts > self.txn_props.start_ts {
-                    MVCC_CONFLICT_COUNTER.prewrite_write_conflict.inc();
-                    self.write_conflict_error(&write, commit_ts)?;
+                match self.txn_props.kind {
+                    TransactionKind::Optimistic(_) => {
+                        if commit_ts > self.txn_props.start_ts {
+                            MVCC_CONFLICT_COUNTER.prewrite_write_conflict.inc();
+                            self.write_conflict_error(&write, commit_ts)?;
+                        }
+                    }
+                    TransactionKind::Pessimistic(for_update_ts) => {
+                        if commit_ts > for_update_ts {
+                            return Err(ErrorInner::PessimisticLockNotFound {
+                                start_ts: self.txn_props.start_ts,
+                                key: self.key.clone().into_raw()?,
+                            }
+                            .into());
+                        }
+                    }
                 }
                 // If there's a write record whose commit_ts equals to our start ts, the current
                 // transaction is ok to continue, unless the record means that the current
@@ -1352,7 +1365,7 @@ pub mod tests {
         must_commit(&engine, b"k1", 10, 20);
         must_commit(&engine, b"k2", 10, 20);
 
-        // This is a re-sent prewrite. It should report a WriteConflict. In production, the caller
+        // This is a re-sent prewrite. It should report a PessimisticLockNotFound. In production, the caller
         // will need to check if the current transaction is already committed before, in order to
         // provide the idempotency.
         let err = must_retry_pessimistic_prewrite_put_err(
@@ -1366,7 +1379,10 @@ pub mod tests {
             false,
             0,
         );
-        assert!(matches!(err, Error(box ErrorInner::WriteConflict { .. })));
+        assert!(matches!(
+            err,
+            Error(box ErrorInner::PessimisticLockNotFound { .. })
+        ));
         // Commit repeatedly, these operations should have no effect.
         must_commit(&engine, b"k1", 10, 25);
         must_commit(&engine, b"k2", 10, 25);
@@ -1382,7 +1398,7 @@ pub mod tests {
         must_commit(&engine, b"k2", 35, 40);
 
         // A retrying non-pessimistic-lock prewrite request should not skip constraint checks.
-        // It reports a WriteConflict.
+        // It reports a PessimisticLockNotFound.
         let err = must_retry_pessimistic_prewrite_put_err(
             &engine,
             b"k2",
@@ -1394,13 +1410,19 @@ pub mod tests {
             false,
             0,
         );
-        assert!(matches!(err, Error(box ErrorInner::WriteConflict { .. })));
+        assert!(matches!(
+            err,
+            Error(box ErrorInner::PessimisticLockNotFound { .. })
+        ));
         must_unlocked(&engine, b"k2");
 
         let err = must_retry_pessimistic_prewrite_put_err(
             &engine, b"k2", b"v2", b"k1", &None, 10, 10, false, 0,
         );
-        assert!(matches!(err, Error(box ErrorInner::WriteConflict { .. })));
+        assert!(matches!(
+            err,
+            Error(box ErrorInner::PessimisticLockNotFound { .. })
+        ));
         must_unlocked(&engine, b"k2");
         // Committing still does nothing.
         must_commit(&engine, b"k2", 10, 25);
@@ -1408,7 +1430,10 @@ pub mod tests {
         let err = must_retry_pessimistic_prewrite_put_err(
             &engine, b"k2", b"v2", b"k1", &None, 11, 11, false, 0,
         );
-        assert!(matches!(err, Error(box ErrorInner::WriteConflict { .. })));
+        assert!(matches!(
+            err,
+            Error(box ErrorInner::PessimisticLockNotFound { .. })
+        ));
         must_unlocked(&engine, b"k2");
         // However conflict still won't be checked if there's a non-retry request arriving.
         must_prewrite_put_impl(
@@ -1417,10 +1442,10 @@ pub mod tests {
             b"v2",
             b"k1",
             &None,
-            10.into(),
+            12.into(),
             false,
             100,
-            10.into(),
+            12.into(),
             1,
             15.into(),
             TimeStamp::default(),
@@ -1428,7 +1453,29 @@ pub mod tests {
             kvproto::kvrpcpb::Assertion::None,
             kvproto::kvrpcpb::AssertionLevel::Off,
         );
-        must_locked(&engine, b"k2", 10);
+        must_locked(&engine, b"k2", 12);
+        must_rollback(&engine, b"k2", 12, false);
+
+        // And conflict check is according to the for_update_ts for pessimistic prewrite.
+        // So, it will not report error if for_update_ts is large enough.
+        must_prewrite_put_impl(
+            &engine,
+            b"k2",
+            b"v2",
+            b"k1",
+            &None,
+            13.into(),
+            false,
+            100,
+            55.into(),
+            1,
+            60.into(),
+            TimeStamp::default(),
+            true,
+            kvproto::kvrpcpb::Assertion::None,
+            kvproto::kvrpcpb::AssertionLevel::Off,
+        );
+        must_locked(&engine, b"k2", 13);
     }
 
     #[test]

--- a/src/storage/txn/commands/prewrite.rs
+++ b/src/storage/txn/commands/prewrite.rs
@@ -1961,12 +1961,12 @@ mod tests {
         // Committing still does nothing.
         must_commit(&engine, b"k2", 10, 25);
         // Try a different txn start ts (which haven't been successfully committed before).
-        // It should report a WriteConflict.
+        // It should report a PessimisticLockNotFound.
         let err = prewrite_with_retry_flag(b"k2", b"v2", b"k1", None, 11, false, true).unwrap_err();
         assert!(matches!(
             err,
             Error(box ErrorInner::Mvcc(MvccError(
-                box MvccErrorInner::WriteConflict { .. }
+                box MvccErrorInner::PessimisticLockNotFound { .. }
             )))
         ));
         must_unlocked(&engine, b"k2");
@@ -2189,7 +2189,7 @@ mod tests {
         );
         assert!(matches!(
             err,
-            MvccError(box MvccErrorInner::WriteConflict { .. })
+            MvccError(box MvccErrorInner::PessimisticLockNotFound { .. })
         ));
     }
 

